File: frontend/src/services/syncService.ts
/**
 * Sync Service for interacting with the Sync API endpoints.
 */

import apiService from './apiService';

// Endpoints
const ENDPOINTS = {
  SYNC: '/sync',
  USER_SYNCS: '/sync/user',
  SYNC_BY_STATE: '/sync/state',
};

export interface Sync {
  id: number;
  user_id: number;
  discussion_name: string;
  state: string;
  created_at: string;
  updated_at: string;
}

export interface SyncCreate {
  user_id: number;
  discussion_name: string;
  state: string;
  dialog_id?: number;
}

export interface SyncUpdate {
  discussion_name?: string;
  state?: string;
}

/**
 * Service to interact with Sync API endpoints.
 */
const syncService = {
  /**
   * Create a new sync.
   * 
   * @param syncData - Data for creating a new sync
   * @returns Promise resolving to the created sync
   */
  async createSync(syncData: SyncCreate): Promise<Sync> {
    return apiService.post<Sync>(ENDPOINTS.SYNC, syncData);
  },

  /**
   * Get all syncs.
   * 
   * @returns Promise resolving to a list of syncs
   */
  async getAllSyncs(): Promise<Sync[]> {
    return apiService.get<Sync[]>(ENDPOINTS.SYNC);
  },

  /**
   * Get syncs for a specific user.
   * 
   * @param userId - ID of the user
   * @returns Promise resolving to a list of syncs
   */
  async getUserSyncs(userId: number): Promise<Sync[]> {
    return apiService.get<Sync[]>(`${ENDPOINTS.USER_SYNCS}/${userId}`);
  },

  /**
   * Get syncs with a specific state.
   * 
   * @param state - State of the syncs to retrieve
   * @returns Promise resolving to a list of syncs
   */
  async getSyncsByState(state: string): Promise<Sync[]> {
    return apiService.get<Sync[]>(`${ENDPOINTS.SYNC_BY_STATE}/${state}`);
  },

  /**
   * Get a specific sync by ID.
   * 
   * @param syncId - ID of the sync to retrieve
   * @returns Promise resolving to the sync
   */
  async getSync(syncId: number): Promise<Sync> {
    return apiService.get<Sync>(`${ENDPOINTS.SYNC}/${syncId}`);
  },

  /**
   * Update a sync by ID.
   * 
   * @param syncId - ID of the sync to update
   * @param syncData - Data to update in the sync
   * @returns Promise resolving to the updated sync
   */
  async updateSync(syncId: number, syncData: SyncUpdate): Promise<Sync> {
    return apiService.post<Sync>(`${ENDPOINTS.SYNC}/${syncId}`, syncData);
  },

  /**
   * Update a sync's state.
   * 
   * @param syncId - ID of the sync to update
   * @param state - New state for the sync
   * @returns Promise resolving to the updated sync
   */
  async updateSyncState(syncId: number, state: string): Promise<Sync> {
    return apiService.put<Sync>(`${ENDPOINTS.SYNC}/${syncId}/state/${state}`, {});
  },

  /**
   * Delete a sync by ID.
   * 
   * @param syncId - ID of the sync to delete
   * @returns Promise resolving to a success status
   */
  async deleteSync(syncId: number): Promise<{ detail: string }> {
    return apiService.delete<{ detail: string }>(`${ENDPOINTS.SYNC}/${syncId}`);
  },
};

export default syncService; 

File: backend/app/services/sync_service.py
"""
Service for Sync operations.
"""

from datetime import datetime, UTC
from sqlalchemy.orm import Session
from typing import List, Optional, Dict, Any, Tuple
from sqlalchemy import select

from app.models.sync import Sync, SyncState
from app.models.telegram_data import User

class SyncService:
    """
    Service for sync operations.
    """
    
    def __init__(self, db: Session):
        self.db = db
    
    def get_sync(self, sync_id: int) -> Optional[Sync]:
        """Get a sync by ID."""
        return self.db.get(Sync, sync_id)
    
    def create_sync(self, user_id: int, discussion_name: str, state: SyncState = SyncState.ACTIVE) -> Sync:
        """Create a new sync."""
        sync = Sync(
            user_id=user_id,
            discussion_name=discussion_name,
            state=state
        )
        self.db.add(sync)
        self.db.commit()
        self.db.refresh(sync)
        return sync
    
    def update_sync(self, sync_id: int, **kwargs) -> Tuple[bool, Optional[Sync]]:
        """
        Update a sync by ID.
        
        Args:
            sync_id: The ID of the sync to update
            **kwargs: Fields to update (discussion_name, state)
            
        Returns:
            Tuple of (success, updated_sync)
        """
        sync = self.get_sync(sync_id)
        if not sync:
            return False, None
        
        # Update the provided fields
        for key, value in kwargs.items():
            if hasattr(sync, key) and key not in ['id', 'user_id', 'created_at']:
                setattr(sync, key, value)
        
        self.db.commit()
        self.db.refresh(sync)
        return True, sync
    
    def delete_sync(self, sync_id: int) -> bool:
        """Delete a sync by ID."""
        sync = self.get_sync(sync_id)
        if sync:
            self.db.delete(sync)
            self.db.commit()
            return True
        return False
    
    def get_all_syncs(self) -> List[Sync]:
        """Get all syncs."""
        return self.db.execute(select(Sync)).scalars().all()
    
    def get_user_syncs(self, user_id: int) -> List[Sync]:
        """Get all syncs for a specific user."""
        return self.db.execute(
            select(Sync).where(Sync.user_id == user_id)
        ).scalars().all()
    
    def get_syncs_by_state(self, state: SyncState) -> List[Sync]:
        """Get all syncs with a specific state."""
        return self.db.execute(
            select(Sync).where(Sync.state == state)
        ).scalars().all()
    
    def update_sync_state(self, sync_id: int, state: SyncState) -> Tuple[bool, Optional[Sync]]:
        """
        Update just the state of a sync.
        
        Args:
            sync_id: The ID of the sync to update
            state: The new state
            
        Returns:
            Tuple of (success, updated_sync)
        """
        return self.update_sync(sync_id, state=state)